icaruscode/PMT/OpReco/ICARUSFlashAssAna_module.cc

```diff
@@ -0,0 +1,457 @@
+    ICARUSFlashAssAna(ICARUSFlashAssAna&&) = delete;
+    ICARUSFlashAssAna& operator=(ICARUSFlashAssAna const&) = delete;
+    ICARUSFlashAssAna& operator=(ICARUSFlashAssAna&&) = delete;
+
+    void analyze(art::Event const& e) override;
+
+    void beginJob() override;
+    void endJob() override;
+
+    int getSideByChannel( const int channel );
+
+    void processOpHits( std::vector<art::Ptr<recob::OpHit>> const &ophits, 
+                        int &multiplicity_left, int &multiplicity_right, 
+                        float &sum_pe_left, float &sum_pe_right, 
+                        float *xyz, TTree *ophittree   ); 
+
+    static std::string_view firstLine(std::string const& s, const char* endl = "\r");
+
+  private:
+
+    art::InputTag fTriggerLabel;
+    std::vector<art::InputTag> fOpHitLabel;
+    std::vector<art::InputTag> fFlashLabels;
+    float fPEOpHitThreshold;
+
+
+    TTree *fEventTree;
+    std::vector<TTree*> fOpFlashTrees;
+    std::vector<TTree*> fOpHitTrees;
+
+    int m_run;
+    int m_event;
+    int m_timestamp;
+    //int m_nflashes;
+    //int m_nophit;
+    float m_beam_gate_start=-99999;
+    float m_beam_gate_width=-99999;
+    int m_beam_type=-1;
+    int m_flash_id;
+    int m_multiplicity;
+    int m_multiplicity_left;
+    int m_multiplicity_right;
+    float m_sum_pe;
+    float m_sum_pe_left;
+    float m_sum_pe_right;
+    float m_flash_time;
+    //float m_flash_x;
+    //float m_flash_width_x;
+    float m_flash_y;
+    float m_flash_width_y;
+    float m_flash_z;
+    float m_flash_width_z;
+
+    int m_channel_id;
+    float m_integral; // in ADC x tick
+    float m_amplitude; // in ADC
+    float m_start_time;
+    float m_width;
+    float m_abs_start_time;
+    float m_pe;
+    float m_fast_to_total;
+
+    std::vector<float> m_pmt_x;
+    std::vector<float> m_pmt_y;
+    std::vector<float> m_pmt_z;
+
+};
+
+
+opana::ICARUSFlashAssAna::ICARUSFlashAssAna(Parameters const& config)
+  : EDAnalyzer(config)
+  , fTriggerLabel( config().TriggerLabel() )
+  , fOpHitLabel( config().OpHitLabels() )
+  , fFlashLabels( config().FlashLabels() )
+  , fPEOpHitThreshold( config().PEOpHitThreshold() )
+{ }
+
+
+void opana::ICARUSFlashAssAna::beginJob() {
+
+  art::ServiceHandle<art::TFileService const> tfs;
+
+  TTree* fGeoTree = tfs->make<TTree>("geotree", "geometry information" );
+  fGeoTree->Branch("pmt_x",&m_pmt_x);
+  fGeoTree->Branch("pmt_y",&m_pmt_y);
+  fGeoTree->Branch("pmt_z",&m_pmt_z);
+  
+  auto const geop = lar::providerFrom<geo::Geometry>();
+
+  double PMTxyz[3];
+  for(size_t opch=0; opch<geop->NOpChannels(); ++opch) {
+
+    geop->OpDetGeoFromOpChannel(opch).GetCenter(PMTxyz);
+
+    //std::cout << PMTxyz[0] << " " << PMTxyz[1] << " " << PMTxyz[2] << std::endl;
+
+    m_pmt_x.push_back(PMTxyz[0]);
+    m_pmt_y.push_back(PMTxyz[1]);
+    m_pmt_z.push_back(PMTxyz[2]);
+
+  }
+
+  fGeoTree->Fill();
+
+  fEventTree = tfs->make<TTree>("eventstree", "higher level information on the event" );
```

Comment:
Could this have a better name like "pmtTree"? Since it's just optical information.